contracts/Stage3/Staking.sol

```diff
@@ -1,78 +1,81 @@
-    uint256 public immutable windowStart;     // the start time for the 'stake'
-    uint256 public immutable windowEnd;       // the end time for the 'stake'
-    uint256 public immutable maturity;        // the maturity date
-    uint256 public immutable initialised;     // the time we initialised the contract
-    uint256 public immutable minTST;          // the min amount of tst we want to allow to bond
-
-    uint256 private _tokenId;
-    uint256 public SEURO_ALLOCATED;           // the amount of seuro allocated, inc rewards
-    bool public active;                       // active or not, needs to be set manually
-    bool public catastrophic;                 // in the event of a catastrophy, let users withdraw
+    uint256 public immutable SI_RATE;   // simple interest rate for the bond (factor of 10 ** 5)
+    uint256 public immutable minTST;    // the allowed minimum amount of TST to bond
 
     mapping(address => Position) private _positions;
 
     struct Position { uint96 nonce; uint256 tokenId; bool open; uint256 stake; uint256 reward; }
 
-    constructor(string memory _name, string memory _symbol, uint256 _start, uint256 _end, uint256 _maturity, address _TST_ADDRESS, address _SEURO_ADDRESS, uint256 _SEUROTST, uint256 _INTEREST) ERC721(_name, _symbol) {
-        SEUROTST = _SEUROTST;
-        INTEREST = _INTEREST;
-        TST_ADDRESS = _TST_ADDRESS;
-        SEURO_ADDRESS = _SEURO_ADDRESS;
+    constructor(string memory _name, string memory _symbol, uint256 _start, uint256 _end, uint256 _maturity, address _gatewayAddress, address _standardAddress, address _seuroAddress, uint256 _si) ERC721(_name, _symbol) {
+        tokenGateway = StandardTokenGateway(_gatewayAddress);
+        SI_RATE = _si;
+        TST_ADDRESS = _standardAddress;
+        SEURO_ADDRESS = _seuroAddress;
         windowStart = _start;
         windowEnd = _end;
         maturity = _maturity;
-        initialised = block.timestamp;
-
-        // TODO variable
+        initialisedAt = block.timestamp;
         minTST = 1 ether;
     }
 
     function activate() external onlyOwner { require(active == false, "err-already-active"); active = true; }
 
     function disable() external onlyOwner { require(active, "err-not-active"); active = false; }
 
-    // reward works out the amount of seuro given to the user based on the
-    // amount of TST they first put in.
-    function calculateReward(uint256 _amount) public view returns (uint256 reward) {
-        uint256 rewardTST = (_amount * INTEREST) / 100_000;
-        reward = (rewardTST * SEUROTST) / 100_000;
+    // calculates the reward in SEURO based in the input of amount of TSTs
+    function calculateReward(uint256 _amountStandard) public view returns (uint256 reward) {
+        (uint256 tokenPrice, bool inverted) = tokenGateway.getSeuroStandardTokenPrice();
+        return (SimpleInterest.convert(_amountStandard, tokenPrice, inverted) * SI_RATE) / RATE_FACTOR;
```

Comment:
i've just realised that you are using the exact same method here to convert TST to sEURO as you are [here](https://github.com/the-standard/ibco/blob/staking-refinement/contracts/Stage2/BondStorage.sol#L100) and [here](https://github.com/the-standard/ibco/blob/staking-refinement/contracts/Stage2/BondStorage.sol#L150) to convert sEURO into TST

Reply:
ok, that's a good point. here we can make it more general with a library 👍.

Reply:
i more mean ... this is a bug where the staking reward is possibly going to be 400x higher than it should be

Reply:
really good bug find, @ewansheldon. happy for you to commit this critical bug fix in this PR now and adapt the tests :). you are free to create better interest rate calculator lib too if you want